src/transformStateWithClones.js

```diff
@@ -6,8 +6,42 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    const { type, extraData } = action;
+
+    switch (type) {
+      case 'clear':
+        stateCopy = {};
+        break;
+
+      case 'addProperties':
+        Object.assign(stateCopy, extraData);
+        break;
+
+      case 'removeProperties':
+        removeKey(action, stateCopy);
+        break;
+
+      default: throw new Error('Unexpected action.type value');
+    }
+
+    stateHistory.push({ ...stateCopy });
+  }
+
+  return stateHistory;
 }
 
+const removeKey = function(actionObject, object) {
+  const { keysToRemove } = actionObject;
+
+  for (const keyToRemove of keysToRemove) {
+    delete object[keyToRemove];
+  }
+};
```

Comment:
it is now unclear what exactly this function does.
- It is called `removeKey`, but it basically removes key**s**
- it accepts two args, but it is totally not clear from their name what exactly they do

It is considered to be a bad practice to add the data type to the name of the variable, as it does not give us any info. That's why it is not clear which args the function accepts